tempodb/encoding/vparquet/block_search.go

```diff
@@ -317,14 +422,121 @@ func rawToResults(ctx context.Context, pf *parquet.File, rgs []parquet.RowGroup,
 	return results
 }
 
-type parquetSearchMetrics struct {
-	// TODO:  this isn't accurate, figure out a good way to measure this
-	//pTraceID *pq.InstrumentedPredicate
+// searchStandardTagValues searches a parquet file for "standard" tags. i.e. tags that don't have unique
+// columns and are contained in labelMappings
+func searchStandardTagValues(ctx context.Context, tag string, pf *parquet.File, cb common.TagCallback) error {
+	rgs := pf.RowGroups()
+	makeIter := makeIterFunc(ctx, rgs, pf)
+
+	keyPred := pq.NewStringInPredicate([]string{tag})
+
+	iter := pq.NewJoinIterator(DefinitionLevelResourceAttrs, []pq.Iterator{
+		makeIter(FieldResourceAttrKey, keyPred, "keys"),
+		makeIter(FieldResourceAttrVal, nil, "values"),
+	}, nil)
+	for {
+		match := iter.Next()
+		if match == nil {
+			break
+		}
+		m := match.ToMap()
+		for _, s := range m["values"] {
+			cb(s.String())
+		}
+	}
+	iter.Close()
+
+	iter = pq.NewJoinIterator(DefinitionLevelResourceSpansILSSpan, []pq.Iterator{
+		makeIter(FieldSpanAttrKey, keyPred, "keys"),
+		makeIter(FieldSpanAttrVal, nil, "values"),
+	}, nil)
+	for {
+		match := iter.Next()
+		if match == nil {
+			break
+		}
+		m := match.ToMap()
+		for _, s := range m["values"] {
+			cb(s.String())
+		}
+	}
+	iter.Close()
+
+	return nil
+}
+
+// searchSpecialTagValues searches a parquet file for all values for the provided column. It first attempts
+// to only pull all values from the column's dictionary. If this fails it falls back to scanning the entire path.
+func searchSpecialTagValues(ctx context.Context, column string, pf *parquet.File, cb common.TagCallback) error {
```

Comment:
Does it make sense build this on top of the existing iterator?  A predicate could be used to collect the distinct string values and still make use of the dictionary when present. The logic would be something like:

* KeepColumnChunk: return true
* KeepColumnPage: 
  * if dictionary present: read it, return false (done with this page)
  * else: return true (must read raw values)
* KeepValue: read string, return false

Reply:
Good call. I swapped to a new predicate `reportValuesPredicate` that works as described. Since it's not a "real" predicate I left it in the `vparquet` folder, but if you think it belongs in `parquetquery` I'd be glad to move it.